File: notificationlistener/src/main/java/com/llw/notify/NotifyHelper.java
package com.llw.notify;

import android.service.notification.StatusBarNotification;

/**
 * @author llw
 * @description NotifyHelper
 * @date 2021/8/12 11:27
 */
public class NotifyHelper {
    private static volatile NotifyHelper instance;
    private NotifyListener notifyListener;

    public static NotifyHelper getInstance() {
        if (instance == null) {
            synchronized (NotifyHelper.class) {
                if (instance == null) {
                    instance = new NotifyHelper();
                }
            }
        }
        return instance;
    }

    /**
     * 设置回调方法
     *
     * @param notifyListener 通知监听
     */
    public void setNotifyListener(NotifyListener notifyListener) {
        this.notifyListener = notifyListener;
    }

    /**
     * 收到消息
     *
     * @param sbn 状态栏通知
     */
    public void onReceive(StatusBarNotification sbn) {
        if (notifyListener != null) {
            notifyListener.onReceiveMessage(sbn);
        }
    }

    /**
     * 移除消息
     *
     * @param sbn 状态栏通知
     */
    public void onRemoved(StatusBarNotification sbn) {
        if (notifyListener != null) {
            notifyListener.onRemovedMessage(sbn);
        }
    }
}

File: notificationlistener/src/main/java/idv/neo/service/notificationlistener/NotifyHelper.kt
package idv.neo.service.notificationlistener

import android.service.notification.StatusBarNotification
import android.util.Log

class NotifyHelper private constructor() {
    private var notifyListener: NotifyListener? = null

    fun onReceive(sbn: StatusBarNotification) {
        notifyListener?.onReceiveMessage(sbn)
    }

    fun onRemoved(sbn: StatusBarNotification) {
        notifyListener?.onRemovedMessage(sbn)
    }

    fun setNotifyListener(notifyListener: NotifyListener?) {
        this.notifyListener = notifyListener
    }

    companion object {
        @Volatile
        private var INSTANCE: NotifyHelper? = null
//        @JvmStatic
//        var instance: NotifyHelper? = null
//            get() {
//                synchronized(NotifyHelper::class.java) {
//                    if (field == null) {
//                        field = NotifyHelper()
//                    }
//                }
//                return field
//            }
//            private set

//        @JvmStatic
//        val instance: NotifyHelper by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
//            NotifyHelper()
//        }
        @JvmStatic
        fun getInstance(): NotifyHelper {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: run {
                    val newInstance = NotifyHelper()
                    //For debug android:process
//                    Log.e("NotifyHelper_Debug", " 創建新實例 : ${newInstance.hashCode()} | " +
//                            "進程ID: ${android.os.Process.myPid()} | " +
//                            "ClassLoader: ${newInstance.javaClass.classLoader}")
                    newInstance
                }.also { INSTANCE = it }
            }
        }
    }

    init {
        // 防禦性檢查：確保不會通過反射創建多個實例
        if (INSTANCE != null) {
            throw IllegalStateException("已存在實例: ${INSTANCE.hashCode()}!" +
                    " 禁止通過反射創建新實例: ${this.hashCode()}")
        }
    }
}
